pkg/bootstrap/bootstrap_test.go

```diff
@@ -0,0 +1,476 @@
+				{
+					expectedErr:  false,
+					expectedBool: true,
+				},
+			},
+		},
+		{
+			name: "GitRepository - sequence of status updates before ready",
+			obj: &sourcev1.GitRepository{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       sourcev1.GitRepositoryKind,
+					APIVersion: sourcev1.GroupVersion.String(),
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:       "flux-system",
+					Namespace:  "flux-system",
+					Generation: 1,
+				},
+			},
+			statuses: []updateStatus{
+				{
+					// observed gen different
+					statusFn: func(o client.Object) {
+						gitRepo := o.(*sourcev1.GitRepository)
+						gitRepo.Status = sourcev1.GitRepositoryStatus{
+							ObservedGeneration: -1,
+						}
+					},
+				},
+				{
+					// ready failing
+					statusFn: func(o client.Object) {
+						gitRepo := o.(*sourcev1.GitRepository)
+						gitRepo.Status = sourcev1.GitRepositoryStatus{
+							ObservedGeneration: 1,
+							Conditions: []metav1.Condition{
+								{Type: meta.ReadyCondition, Status: metav1.ConditionFalse, ObservedGeneration: 1, Reason: "Not Ready", Message: "Transient connection issue"},
+							},
+						}
+					},
+					expectedErr: true,
+				},
+				{
+					// updated to a different revision
+					statusFn: func(o client.Object) {
+						gitRepo := o.(*sourcev1.GitRepository)
+						gitRepo.Status = sourcev1.GitRepositoryStatus{
+							ObservedGeneration: 1,
+							Conditions: []metav1.Condition{
+								{Type: meta.ReadyCondition, Status: metav1.ConditionTrue, ObservedGeneration: 1, Reason: "Readyyy", Message: "Cloned successfully"},
+							},
+							Artifact: &sourcev1.Artifact{
+								Revision: "wrong rev",
+							},
+						}
+					},
+				},
+				{
+					// updated to the expected revision
+					statusFn: func(o client.Object) {
+						gitRepo := o.(*sourcev1.GitRepository)
+						gitRepo.Status = sourcev1.GitRepositoryStatus{
+							ObservedGeneration: 1,
+							Conditions: []metav1.Condition{
+								{Type: meta.ReadyCondition, Status: metav1.ConditionTrue, ObservedGeneration: 1, Reason: "Readyyy", Message: "Cloned successfully"},
+							},
+							Artifact: &sourcev1.Artifact{
+								Revision: expectedRev,
+							},
+						}
+					},
+					expectedBool: true,
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+			builder := fake.NewClientBuilder().WithScheme(getScheme())
+			builder.WithObjects(tt.obj)
+
+			kubeClient := builder.Build()
+
+			for _, updates := range tt.statuses {
+				if updates.statusFn != nil {
+					updates.statusFn(tt.obj)
+					g.Expect(kubeClient.Update(context.TODO(), tt.obj)).To(Succeed())
+				}
+
+				waitFunc := objectReconciled(kubeClient, client.ObjectKeyFromObject(tt.obj), tt.obj, expectedRev)
+				got, err := waitFunc(context.TODO())
+				g.Expect(err != nil).To(Equal(updates.expectedErr))
+				g.Expect(got).To(Equal(updates.expectedBool))
+			}
+		})
+	}
+}
+
+func getScheme() *runtime.Scheme {
```

Comment:
The scheme for tests is here https://github.com/fluxcd/flux2/blob/c5b5263b1bd826e792b3e5607857b7cd6d981b62/internal/utils/utils.go#L124
```suggestion
func getScheme() *runtime.Scheme {
```